### PR TITLE
fix the 'FIXME' in code: add a flag that track whether we've already …

### PR DESCRIPTION
**Pull request by LoveULin**
…dispatched

### CODE REVIEW

**Comment by nbougalis** [comment]
Thanks for your contribution. I am assigning reviewers for this.

File: src/ripple/app/ledger/LedgerMaster.h [hunk 1]
```diff
@@ -342,6 +342,8 @@ class LedgerMaster
     int     mPathFindThread {0};    // Pathfinder jobs dispatched
     bool    mPathFindNewRequest {false};
 
+    std::atomic_flag mGotFetchPackThread {ATOMIC_FLAG_INIT}; // GotFetchPack jobs dispatched
```
[the remark this answers is not in the document]
Comment (LoveULin) [reply src/ripple/app/ledger/LedgerMaster.h]: OK. Thank you. I did not try to compile on macOS. And the use of:

`std::atomic_flag mGotFetchPackThread = ATOMIC_FLAG_INIT;`

also seems OK on VS2017 and gcc, I'll change the code to this.

File: src/ripple/app/ledger/impl/LedgerMaster.cpp [hunk 2]
```diff
@@ -1779,13 +1779,16 @@ LedgerMaster::gotFetchPack (
...
     std::uint32_t seq)
 {
-    // FIXME: Calling this function more than once will result in
-    // InboundLedgers::gotFetchPack being called more than once
-    // which is expensive. A flag should track whether we've already dispatched
-
-    app_.getJobQueue().addJob (
-        jtLEDGER_DATA, "gotFetchPack",
-        [&] (Job&) { app_.getInboundLedgers().gotFetchPack(); });
+    if (!mGotFetchPackThread.test_and_set(std::memory_order::memory_order_consume))
```
[the remark this answers is not in the document]
Comment (LoveULin) [reply src/ripple/app/ledger/impl/LedgerMaster.cpp]: Thanks for the review. I really appreciate it. After looking for some extra information about `memory_order`, I think you are right:
>At the very least I believe we need `memory_order_acquire` here. 

As for the use of `memory_order`, I agree with that:
>do not use anything other than std::memory_order_seq_cst (the default) unless (a) you really really know what you are doing, and can prove that the relaxed usage is safe in all cases, and (b) your profiler demonstrates that the data structure and operations you are intending to use the relaxed orderings with are a bottleneck.

My original idea was to just make the code more strict and efficient. I'll fix the `memory_order_consume` to `memory_order_acquire` for the time being. Then should we use the default `memory order` here directly??

File: src/ripple/app/ledger/impl/LedgerMaster.cpp [hunk 3]
```diff
@@ -1779,13 +1779,16 @@ LedgerMaster::gotFetchPack (
...
+    {
+        app_.getJobQueue().addJob (
+            jtLEDGER_DATA, "gotFetchPack",
+            [&] (Job&)
+            {
+                app_.getInboundLedgers().gotFetchPack();
+                mGotFetchPackThread.clear(std::memory_order::memory_order_release);
```
Comment (miguelportilla) [inline src/ripple/app/ledger/impl/LedgerMaster.cpp]: Suggestion: Use `std::memory_order_release` vs `std::memory_order::memory_order_release`

File: src/ripple/app/ledger/impl/LedgerMaster.cpp [hunk 4]
```diff
@@ -1779,7 +1779,7 @@ LedgerMaster::gotFetchPack (
     bool progress,
     std::uint32_t seq)
 {
-    if (!mGotFetchPackThread.test_and_set(std::memory_order::memory_order_consume))
+    if (!mGotFetchPackThread.test_and_set(std::memory_order::memory_order_acquire))
```
Comment (miguelportilla) [inline src/ripple/app/ledger/impl/LedgerMaster.cpp]: Suggestion: Use `std::memory_order_acquire` vs `std::memory_order::memory_order_acquire`

**Comment by nbougalis** [comment]
@LoveULin just so we can set expectations: I don't think this will go into **1.0**; the *develop* branch is in a semi-frozen state and ready to transition from *beta* to *release candidate* so only fixes to included stuff are going in;

Your PR will be integrated in the **1.1.0-b1**.

**Comment by LoveULin** [comment]
OK, I got it. And I have squashed the commits together into one, and reworded the single resulting commit according to your suggestion.

Thank you for your instructions and guidance, really :D

**Comment by JoelKatz** [comment]
There's a race condition here, but it's benign. Imagine if we receive a new fetch pack just as `gotFetchPack` is returning but before `mGotFetchPackThread` is cleared. In that case, we won't dispatch a new job to process the new fetch pack even though it hasn't been processed against any existing inbound ledger jobs. The inbound ledgers will get processed anyway as soon as they receive any data or timeout. So in this case, the race condition is harmless and the value of avoiding duplicate dispatches is much greater.

Thanks for the contribution.

**Comment by JoeLoser** [comment]
@nbougalis think we should try to land this in 1.1.0?

**Comment by nbougalis** [comment]
Merged as 4dcb3c9199fcf46f3c20baed008d990a0ac95b42

**Pull request closed without merging**